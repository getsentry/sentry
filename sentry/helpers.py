import warnings

warnings.warn("The use of 'sentry.helpers' is deprecated. Please use 'sentry.utils' instead.", DeprecationWarning)

from sentry.utils import *
