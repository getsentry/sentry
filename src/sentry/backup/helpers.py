from __future__ import annotations

from datetime import datetime, timedelta, timezone
from enum import Enum
from typing import Generic, NamedTuple, Type, TypeVar

from django.core.serializers.json import DjangoJSONEncoder
from django.db import models

# Django apps we take care to never import or export from.
EXCLUDED_APPS = frozenset(("auth", "contenttypes", "fixtures"))

UTC_0 = timezone(timedelta(hours=0))


class Printer:
    """
    A simplified interface for a terminal CLI input-output interface. The default implementation is
    a no-op.
    """

    def echo(
        self,
        text: str,
        *,
        err: bool = False,
        color: bool | None = None,
    ) -> None:
        pass

    def confirm(
        self,
        text: str,
        *,
        default: bool | None = None,
        err: bool = False,
    ) -> bool:
        return True


class DatetimeSafeDjangoJSONEncoder(DjangoJSONEncoder):
    """
    A wrapper around the default `DjangoJSONEncoder` that always retains milliseconds, even when
    their implicit value is `.000`. This is necessary because the ECMA-262 compatible
    `DjangoJSONEncoder` drops these by default.
    """

    def default(self, obj):
        if isinstance(obj, datetime):
            return obj.astimezone(UTC_0).strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z"
        return super().default(obj)


class Side(Enum):
    """
    Used to identify the "side" in backup operations which perform comparisons between two sets of exports JSONs. The "left" side is usually the older of the two states (ie, "left" is roughly synonymous with "before", and "right" with "after").
    """

    left = 1
    right = 2


T = TypeVar("T")


class Filter(Generic[T]):
    """Specifies a field-based filter when performing an import or export operation. This is an
    allowlist based filtration: models of the given type whose specified field matches ANY of the
    supplied values will be allowed through."""

    model: Type[models.base.Model]
    field: str
    values: set[T]

    def __init__(self, model: Type[models.base.Model], field: str, values: set[T] | None = None):
        self.model = model
        self.field = field
        self.values = values if values is not None else set()


class ImportFlags(NamedTuple):
    """
    Flags that affect how importing a relocation JSON file proceeds.
    """

    # If a username already exists, should we re-use that user, or create a new one with a randomly
    # suffixed username (ex: "some-user" would become "some-user-ad21")
    merge_users: bool = False

    # If a global configuration value `ControlOption`/`Option` (as identified by its unique
    # `key`) or `Relay` (as identified by its unique `relay_id`) already exists, should we overwrite
    # it with the new value, or keep the existing one and discard the incoming value instead?
    overwrite_configs: bool = False

    # A UUID with which to identify this import's `*ImportChunk` database entries. Useful for
    # passing the calling `Relocation` model's UUID to all of the imports it triggered. If this flag
    # is not provided, the import was called in a non-relocation context, like from the `sentry
    # import` CLI command.
    import_uuid: str | None = None
