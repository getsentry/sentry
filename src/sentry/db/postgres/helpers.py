from __future__ import absolute_import

import psycopg2
import six

from django.db.utils import DatabaseError, InterfaceError


def can_reconnect(exc):
    if isinstance(exc, (psycopg2.InterfaceError, InterfaceError)):
        return True
    # elif isinstance(exc, psycopg2.OperationalError):
    #     exc_msg = six.text_type(exc)
    #     if "can't fetch default_isolation_level" in exc_msg:
    #         return True
    #     elif "can't set datestyle to ISO" in exc_msg:
    #         return True
    #     return True
    elif isinstance(exc, DatabaseError):
        exc_msg = six.text_type(exc)
        if "server closed the connection unexpectedly" in exc_msg:
            return True
        elif "client_idle_timeout" in exc_msg:
            return True
    return False
