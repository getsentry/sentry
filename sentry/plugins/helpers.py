"""
sentry.plugins.helpers
~~~~~~~~~~~~~~~~~~~~~~

:copyright: (c) 2010-2012 by the Sentry Team, see AUTHORS for more details.
:license: BSD, see LICENSE for more details.
"""

from sentry.models import ProjectOption, Option, UserOption

__all__ = ('set_option', 'get_option', 'unset_option')

def set_option(key, value, project=None, user=None):
    if user:
        result = UserOption.objects.set_value(user, project, key, value)
    elif project:
        result = ProjectOption.objects.set_value(project, key, value)
    else:
        result = Option.objects.set_value(key, value)
    return result


def get_option(key, project=None, user=None, default=None):
    if user:
        result = UserOption.objects.get_value(user, project, key, default)
    elif project:
        result = ProjectOption.objects.get_value(project, key, default)
    else:
        result = Option.objects.get_value(key, default)
    return result


def unset_option(key, project=None, user=None):
    if user:
        result = UserOption.objects.unset_value(user, project, key)
    elif project:
        result = ProjectOption.objects.unset_value(project, key)
    else:
        result = Option.objects.unset_value(key)

    return result
