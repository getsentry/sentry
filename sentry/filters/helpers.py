"""
sentry.filters.helpers
~~~~~~~~~~~~~~~~~~~~~~

:copyright: (c) 2010-2012 by the Sentry Team, see AUTHORS for more details.
:license: BSD, see LICENSE for more details.
"""

# Widget api is pretty ugly
from __future__ import absolute_import

__all__ = ('get_filters',)

import logging

from sentry.conf import settings
from sentry.plugins import plugins


FILTER_CACHE = {}


def get_filters(model=None, project=None):
    filter_list = []

    # Add builtins (specified with the FILTERS setting)
    for class_path in settings.FILTERS:
        if class_path not in FILTER_CACHE:
            module_name, class_name = class_path.rsplit('.', 1)
            try:
                module = __import__(module_name, {}, {}, class_name)
                cls = getattr(module, class_name)
            except Exception:
                logger = logging.getLogger('sentry.errors.filters')
                logger.exception('Unable to import %s', class_path)
                continue
            FILTER_CACHE[class_path] = cls
        filter_list.append(FILTER_CACHE[class_path])

    # Add plugin-provided filters
    for plugin in plugins.all():
        if not plugin.is_enabled(project):
            continue

        for filter_cls in plugin.get_filters(project):
            if filter_cls not in filter_list:
                filter_list.append(filter_cls)

    # yield all filters which support ``model``
    for filter_cls in filter_list:
        if model and model not in filter_cls.types:
            continue
        yield filter_cls
