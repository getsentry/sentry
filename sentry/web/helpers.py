"""
sentry.web.views
~~~~~~~~~~~~~~~~

:copyright: (c) 2010-2012 by the Sentry Team, see AUTHORS for more details.
:license: BSD, see LICENSE for more details.
"""

import logging

from django.conf import settings as dj_settings
from django.core.urlresolvers import reverse, resolve
from django.db.models import Q
from django.http import HttpResponse
from django.template import loader, RequestContext, Context
from django.utils.datastructures import SortedDict
from django.utils.safestring import mark_safe

from sentry.conf import settings
from sentry.models import Project, View, \
  MEMBER_USER, Option, ProjectOption, Team
from sentry.permissions import can_create_projects, can_create_teams

logger = logging.getLogger('sentry.errors')


def get_project_list(user=None, access=None, hidden=False, key='id'):
    """
    Returns a SortedDict of all projects a user has some level of access to.
    """
    if access is None:
        access = MEMBER_USER

    base_qs = Project.objects
    if not hidden:
        base_qs = base_qs.filter(status=0)

    # Collect kwarg queries to filter on. We can use this to perform a single
    # query to get all of the desired projects ordered by name
    filters = Q()

    # If we're not requesting specific access include all
    # public projects
    if access <= MEMBER_USER:
        filters |= Q(public=True)

    # If the user is authenticated, include their memberships
    if user and user.is_authenticated():
        teams = get_team_list(user, access).values()
        filters |= Q(team__in=teams)

    return SortedDict((getattr(p, key), p)
        for p in base_qs.filter(filters).order_by('name'))


def get_team_list(user, access=None):
    """
    Returns a SortedDict of all teams a user has some level of access to.
    """
    if access is None:
        access = MEMBER_USER

    if not user.is_authenticated():
        return SortedDict()

    qs = Team.objects.filter(
        member_set__user=user,
        member_set__is_active=True,
        member_set__type__lte=access,
    ).order_by('name')
    return SortedDict((p.slug, p) for p in qs)

_LOGIN_URL = None


def get_login_url(reset=False):
    global _LOGIN_URL

    if _LOGIN_URL is None or reset:
        # if LOGIN_URL resolves force login_required to it instead of our own
        # XXX: this must be done as late as possible to avoid idempotent requirements
        try:
            resolve(dj_settings.LOGIN_URL)
        except:
            _LOGIN_URL = settings.LOGIN_URL
        else:
            _LOGIN_URL = dj_settings.LOGIN_URL

        if _LOGIN_URL is None:
            _LOGIN_URL = reverse('sentry-login')
    return _LOGIN_URL


def iter_data(obj):
    for k, v in obj.data.iteritems():
        if k.startswith('_') or k in ['url']:
            continue
        yield k, v


def get_default_context(request, existing_context=None):
    from sentry.plugins import plugins

    context = {
        'HAS_SEARCH': settings.USE_SEARCH,
        'MESSAGES_PER_PAGE': settings.MESSAGES_PER_PAGE,
        'PROJECT_ID': int(settings.PROJECT),
        'VIEWS': list(View.objects.all()),
        'URL_PREFIX': settings.URL_PREFIX,
        'PLUGINS': plugins,
        'USE_JS_CLIENT': settings.USE_JS_CLIENT,
    }

    if request:
        context.update({
            'request': request,
            'can_create_projects': can_create_projects(request.user),
            'can_create_teams': can_create_teams(request.user),
        })
        if not existing_context or 'PROJECT_LIST' not in existing_context:
            context['PROJECT_LIST'] = get_project_list(request.user).values()
        if not existing_context or 'TEAM_LIST' not in existing_context:
            context['TEAM_LIST'] = get_team_list(request.user).values()

    return context


def render_to_string(template, context=None, request=None):
    default_context = get_default_context(request, context)

    if context is None:
        context = default_context
    else:
        context = dict(context)
        context.update(default_context)

    if request:
        context = RequestContext(request, context)
    else:
        context = Context(context)

    return loader.render_to_string(template, context)


def render_to_response(template, context=None, request=None, status=200):
    response = HttpResponse(render_to_string(template, context, request))
    response.status_code = status

    return response


def plugin_config(plugin, project, request):
    """
    Configure the plugin site wide.

    returns a tuple composed of a redirection boolean and the content to
    be displayed.
    """
    NOTSET = object()

    plugin_key = plugin.get_conf_key()
    if project:
        form_class = plugin.project_conf_form
        template = plugin.project_conf_template
    else:
        form_class = plugin.site_conf_form
        template = plugin.site_conf_template

    initials = plugin.get_form_initial(project)
    for field in form_class.base_fields:
        key = '%s:%s' % (plugin_key, field)
        if project:
            value = ProjectOption.objects.get_value(project, key, NOTSET)
        else:
            value = Option.objects.get_value(key, NOTSET)
        if value is not NOTSET:
            initials[field] = value

    form = form_class(
        request.POST or None,
        initial=initials,
        prefix=plugin_key
    )
    if form.is_valid():
        for field, value in form.cleaned_data.iteritems():
            key = '%s:%s' % (plugin_key, field)
            if project:
                ProjectOption.objects.set_value(project, key, value)
            else:
                Option.objects.set_value(key, value)

        return ('redirect', None)

    from django.template.loader import render_to_string
    return ('display', mark_safe(render_to_string(template, {
            'form': form,
            'request': request,
            'plugin_description': plugin.get_description() or '',
        }, context_instance=RequestContext(request))))
