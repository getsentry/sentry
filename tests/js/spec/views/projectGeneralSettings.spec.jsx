import {browserHistory} from 'react-router';
import React from 'react';

import {mount} from 'enzyme';
import ProjectContext from 'app/views/projects/projectContext';
import ProjectGeneralSettings from 'app/views/settings/projectGeneralSettings';
import ProjectsStore from 'app/stores/projectsStore';

import {selectByValue} from '../../helpers/select';

jest.mock('jquery');

describe('projectGeneralSettings', function() {
  let org = TestStubs.Organization();
  let project = TestStubs.ProjectDetails();
  let routerContext;
  let putMock;

  beforeEach(function() {
    sinon.stub(window.location, 'assign');
    routerContext = TestStubs.routerContext([
      {
        router: TestStubs.router({
          params: {
            projectId: project.slug,
            orgId: org.slug,
          },
        }),
      },
    ]);

    MockApiClient.clearMockResponses();
    MockApiClient.addMockResponse({
      url: `/projects/${org.slug}/${project.slug}/`,
      method: 'GET',
      body: project,
    });
    MockApiClient.addMockResponse({
      url: `/projects/${org.slug}/${project.slug}/environments/`,
      method: 'GET',
      body: [],
    });
    MockApiClient.addMockResponse({
      url: `/projects/${org.slug}/${project.slug}/members/`,
      method: 'GET',
      body: [],
    });
  });

  afterEach(function() {
    window.location.assign.restore();
  });

  it('renders form fields', function() {
    let wrapper = mount(
      <ProjectGeneralSettings params={{orgId: org.slug, projectId: project.slug}} />,
      TestStubs.routerContext()
    );

    expect(wrapper.find('Input[name="name"]').prop('value')).toBe('Project Name');
    expect(wrapper.find('Input[name="slug"]').prop('value')).toBe('project-slug');
    expect(wrapper.find('Input[name="subjectPrefix"]').prop('value')).toBe('[my-org]');
    expect(wrapper.find('RangeSlider[name="resolveAge"]').prop('value')).toBe(48);
    expect(wrapper.find('Switch[name="dataScrubber"]').prop('isActive')).toBeFalsy();
    expect(
      wrapper.find('Switch[name="dataScrubberDefaults"]').prop('isActive')
    ).toBeFalsy();
    expect(wrapper.find('Switch[name="scrubIPAddresses"]').prop('isActive')).toBeFalsy();
    expect(wrapper.find('TextArea[name="sensitiveFields"]').prop('value')).toBe(
      'creditcard\nssn'
    );
    expect(wrapper.find('TextArea[name="safeFields"]').prop('value')).toBe(
      'business-email\ncompany'
    );
    expect(wrapper.find('TextArea[name="allowedDomains"]').prop('value')).toBe(
      'example.com\nhttps://example.com'
    );
    expect(wrapper.find('Switch[name="scrapeJavaScript"]').prop('isDisabled')).toBe(
      false
    );
    expect(wrapper.find('Switch[name="scrapeJavaScript"]').prop('isActive')).toBeTruthy();
    expect(wrapper.find('Input[name="securityToken"]').prop('value')).toBe(
      'security-token'
    );
    expect(wrapper.find('Input[name="securityTokenHeader"]').prop('value')).toBe(
      'x-security-header'
    );
    expect(wrapper.find('Switch[name="verifySSL"]').prop('isActive')).toBeTruthy();
  });

  it('disables field when equivalent org setting is true', function() {
    routerContext.context.organization.dataScrubber = true;
    routerContext.context.organization.scrubIPAddresses = false;
    let wrapper = mount(
      <ProjectGeneralSettings params={{orgId: org.slug, projectId: project.slug}} />,
      routerContext
    );
    expect(wrapper.find('Switch[name="scrubIPAddresses"]').prop('isDisabled')).toBe(
      false
    );
    expect(wrapper.find('Switch[name="scrubIPAddresses"]').prop('isActive')).toBeFalsy();
    expect(wrapper.find('Switch[name="dataScrubber"]').prop('isDisabled')).toBe(true);
    expect(wrapper.find('Switch[name="dataScrubber"]').prop('isActive')).toBe(true);
  });

  it('disables scrapeJavaScript when equivalent org setting is false', function() {
    routerContext.context.organization.scrapeJavaScript = false;
    let wrapper = mount(
      <ProjectGeneralSettings params={{orgId: org.slug, projectId: project.slug}} />,
      routerContext
    );
    expect(wrapper.find('Switch[name="scrapeJavaScript"]').prop('isDisabled')).toBe(true);
    expect(wrapper.find('Switch[name="scrapeJavaScript"]').prop('isActive')).toBeFalsy();
  });

  it('project admins can remove project', function() {
    let deleteMock = MockApiClient.addMockResponse({
      url: `/projects/${org.slug}/${project.slug}/`,
      method: 'DELETE',
    });

    let wrapper = mount(
      <ProjectGeneralSettings params={{orgId: org.slug, projectId: project.slug}} />,
      TestStubs.routerContext()
    );

    let removeBtn = wrapper.find('.ref-remove-project').first();

    expect(removeBtn.prop('children')).toBe('Remove Project');

    // Click button
    removeBtn.simulate('click');

    // Confirm Modal
    wrapper.find('Modal Button[priority="danger"]').simulate('click');

    expect(deleteMock).toHaveBeenCalled();
  });

  it('project admins can transfer project', function() {
    let deleteMock = MockApiClient.addMockResponse({
      url: `/projects/${org.slug}/${project.slug}/transfer/`,
      method: 'POST',
    });

    let wrapper = mount(
      <ProjectGeneralSettings params={{orgId: org.slug, projectId: project.slug}} />,
      TestStubs.routerContext()
    );

    let removeBtn = wrapper.find('.ref-transfer-project').first();

    expect(removeBtn.prop('children')).toBe('Transfer Project');

    // Click button
    removeBtn.simulate('click');

    // Confirm Modal
    wrapper
      .find('input[name="email"]')
      .simulate('change', {target: {value: 'billy@sentry.io'}});
    wrapper.find('Modal Button[priority="danger"]').simulate('click');

    expect(deleteMock).toHaveBeenCalledWith(
      `/projects/${org.slug}/${project.slug}/transfer/`,
      expect.objectContaining({
        method: 'POST',
        data: {
          email: 'billy@sentry.io',
        },
      })
    );
  });

  it('displays transfer/remove message for non-admins', function() {
    routerContext.context.organization.access = ['org:read'];
    let wrapper = mount(
      <ProjectGeneralSettings params={{orgId: org.slug, projectId: project.slug}} />,
      routerContext
    );

    expect(wrapper.html()).toContain(
      'You do not have the required permission to remove this project.'
    );
    expect(wrapper.html()).toContain(
      'You do not have the required permission to transfer this project.'
    );
  });

  it('disables the form for users without write permissions', function() {
    routerContext.context.organization.access = ['org:read'];
    let wrapper = mount(
      <ProjectGeneralSettings params={{orgId: org.slug, projectId: project.slug}} />,
      routerContext
    );

    expect(wrapper.find('FormField[disabled=false]')).toHaveLength(0);
    expect(
      wrapper
        .find('Alert')
        .first()
        .text()
    ).toBe(
      'These settings can only be edited by users with the owner, manager, or admin role.'
    );
  });

  it('changing project platform updates ProjectsStore', async function() {
    let params = {orgId: org.slug, projectId: project.slug};
    ProjectsStore.loadInitialData([project]);
    putMock = MockApiClient.addMockResponse({
      url: `/projects/${org.slug}/${project.slug}/`,
      method: 'PUT',
      body: {
        ...project,
        platform: 'javascript',
      },
    });
    let wrapper = mount(
      <ProjectContext orgId={org.slug} projectId={project.slug}>
        <ProjectGeneralSettings
          routes={[]}
          location={routerContext.context.location}
          params={params}
        />
      </ProjectContext>,
      routerContext
    );
    await tick();
    wrapper.update();

    // Change slug to new-slug
    selectByValue(wrapper, 'javascript');

    // Slug does not save on blur
    expect(putMock).toHaveBeenCalled();

    await tick();
    await tick();
    wrapper.update();

    // updates ProjectsStore
    expect(ProjectsStore.itemsById['2'].platform).toBe('javascript');
  });

  it('changing slug updates ProjectsStore', async function() {
    let params = {orgId: org.slug, projectId: project.slug};
    ProjectsStore.loadInitialData([project]);
    putMock = MockApiClient.addMockResponse({
      url: `/projects/${org.slug}/${project.slug}/`,
      method: 'PUT',
      body: {
        ...project,
        slug: 'new-project',
      },
    });
    let wrapper = mount(
      <ProjectContext orgId={org.slug} projectId={project.slug}>
        <ProjectGeneralSettings
          routes={[]}
          location={routerContext.context.location}
          params={params}
        />
      </ProjectContext>,
      routerContext
    );
    await tick();
    wrapper.update();

    // Change slug to new-slug
    wrapper
      .find('input[name="slug"]')
      .simulate('change', {target: {value: 'NEW PROJECT'}})
      .simulate('blur');

    // Slug does not save on blur
    expect(putMock).not.toHaveBeenCalled();
    wrapper.find('SaveButton').simulate('click');

    // fetches new slug
    let newProjectGet = MockApiClient.addMockResponse({
      url: `/projects/${org.slug}/new-project/`,
      method: 'GET',
      body: {...project, slug: 'new-project'},
    });
    let newProjectEnv = MockApiClient.addMockResponse({
      url: `/projects/${org.slug}/new-project/environments/`,
      method: 'GET',
      body: [],
    });
    let newProjectMembers = MockApiClient.addMockResponse({
      url: `/projects/${org.slug}/new-project/members/`,
      method: 'GET',
      body: [],
    });

    await tick();
    // :(
    await tick();
    wrapper.update();
    // updates ProjectsStore
    expect(ProjectsStore.itemsById['2'].slug).toBe('new-project');
    expect(browserHistory.replace).toHaveBeenCalled();
    expect(wrapper.find('Input[name="slug"]').prop('value')).toBe('new-project');

    wrapper.setProps({
      projectId: 'new-project',
    });
    await tick();
    wrapper.update();
    expect(newProjectGet).toHaveBeenCalled();
    expect(newProjectEnv).toHaveBeenCalled();
    expect(newProjectMembers).toHaveBeenCalled();
  });

  describe('Non-"save on blur" Field', function() {
    let wrapper;

    beforeEach(function() {
      let params = {orgId: org.slug, projectId: project.slug};
      ProjectsStore.loadInitialData([project]);
      putMock = MockApiClient.addMockResponse({
        url: `/projects/${org.slug}/${project.slug}/`,
        method: 'PUT',
        body: {
          ...project,
          slug: 'new-project',
        },
      });
      wrapper = mount(
        <ProjectContext orgId={org.slug} projectId={project.slug}>
          <ProjectGeneralSettings
            routes={[]}
            location={routerContext.context.location}
            params={params}
          />
        </ProjectContext>,
        routerContext
      );
    });

    it('can cancel unsaved changes for a field', async function() {
      await tick();
      wrapper.update();
      // Initially does not have "Cancel" button
      expect(wrapper.find('MessageAndActions CancelButton')).toHaveLength(0);
      // Has initial value
      expect(wrapper.find('input[name="resolveAge"]').prop('value')).toBe(19);

      // Change value
      wrapper
        .find('input[name="resolveAge"]')
        .simulate('input', {target: {value: 12}})
        .simulate('mouseUp');

      // Has updated value
      expect(wrapper.find('input[name="resolveAge"]').prop('value')).toBe(12);
      // Has "Cancel" button visible
      expect(wrapper.find('MessageAndActions CancelButton')).toHaveLength(1);

      // Click cancel
      wrapper.find('MessageAndActions CancelButton').simulate('click');
      // Cancel row should disappear
      expect(wrapper.find('MessageAndActions CancelButton')).toHaveLength(0);
      // Value should be reverted
      expect(wrapper.find('input[name="resolveAge"]').prop('value')).toBe(19);
      // PUT should not be called
      expect(putMock).not.toHaveBeenCalled();
    });

    it('saves when value is changed and "Save" clicked', async function() {
      await tick();
      wrapper.update();
      // Initially does not have "Save" button
      expect(wrapper.find('MessageAndActions SaveButton')).toHaveLength(0);

      // Change value
      wrapper
        .find('input[name="resolveAge"]')
        .simulate('input', {target: {value: 12}})
        .simulate('mouseUp');

      // Has "Save" button visible
      expect(wrapper.find('MessageAndActions SaveButton')).toHaveLength(1);

      // Should not have put mock called yet
      expect(putMock).not.toHaveBeenCalled();

      // Click "Save"
      wrapper.find('MessageAndActions SaveButton').simulate('click');
      // API endpoint should have been called
      expect(putMock).toHaveBeenCalledWith(
        expect.anything(),
        expect.objectContaining({
          data: {
            resolveAge: 12,
          },
        })
      );

      // Should hide "Save" button after saving
      await tick();
      wrapper.update();
      expect(wrapper.find('MessageAndActions SaveButton')).toHaveLength(0);
    });
  });
});
