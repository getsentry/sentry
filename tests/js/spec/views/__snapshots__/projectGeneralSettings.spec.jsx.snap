// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`projectGeneralSettings renders 1`] = `
<SideEffect(DocumentTitle)
  title="Sentry"
>
  <div>
    <SettingsPageHeading
      title="Project Settings"
    />
    <ApiForm
      apiEndpoint="/projects/org-slug/project-slug/"
      apiMethod="PUT"
      cancelLabel="Cancel"
      className="form-stacked"
      errorMessage="Unable to save your changes. Please ensure all fields are valid and try again."
      footerClass="form-actions align-right"
      hideErrors={false}
      initialData={
        Object {
          "allowedDomains": undefined,
          "dataScrubber": undefined,
          "dataScrubberDefaults": undefined,
          "defaultEnvironment": undefined,
          "name": "Project Name",
          "resolveAge": undefined,
          "safeFields": undefined,
          "scrapeJavaScript": undefined,
          "scrubIPAddresses": undefined,
          "securityHeader": undefined,
          "securityToken": undefined,
          "securityTokenHeader": undefined,
          "sensitiveFields": undefined,
          "slug": "project-slug",
          "subjectPrefix": undefined,
          "team": undefined,
          "verifySSL": undefined,
        }
      }
      onSubmitSuccess={[Function]}
      requireChanges={false}
      resetOnError={false}
      submitDisabled={false}
      submitLabel="Save Changes"
    >
      <div
        className="box"
      >
        <div
          className="box-header"
        >
          <h3>
            Project Details
          </h3>
        </div>
        <div
          className="box-content with-padding"
        >
          <TextField
            disabled={false}
            hideErrorMessage={false}
            label="Project Name"
            name="name"
            placeholder="e.g. My Service Name"
            required={true}
          />
          <TextField
            disabled={false}
            help="A unique ID used to identify this project."
            hideErrorMessage={false}
            label="Short name"
            name="slug"
            required={true}
          />
        </div>
      </div>
      <div
        className="box"
      >
        <div
          className="box-header"
        >
          <h3>
            Email
          </h3>
        </div>
        <div
          className="box-content with-padding"
        >
          <TextField
            disabled={false}
            help="Choose a custom prefix for emails from this project."
            hideErrorMessage={false}
            label="Subject prefix"
            name="subjectPrefix"
            required={false}
          />
        </div>
      </div>
      <div
        className="box"
      >
        <div
          className="box-header"
        >
          <h3>
            Event Settings
          </h3>
        </div>
        <div
          className="box-content with-padding"
        >
          <TextField
            disabled={false}
            help="The default selected environment when viewing issues."
            hideErrorMessage={false}
            label="Default environment"
            name="defaultEnvironment"
            placeholder="e.g. production"
            required={false}
          />
          <RangeField
            allowedValues={
              Array [
                0,
                1,
                2,
                3,
                4,
                5,
                6,
                7,
                8,
                9,
                10,
                11,
                12,
                15,
                18,
                21,
                24,
                30,
                36,
                48,
                72,
                96,
                120,
                144,
                168,
                192,
                216,
                240,
                264,
                288,
                312,
                336,
                360,
                384,
                408,
                432,
                456,
                480,
                504,
                528,
                552,
                576,
                600,
                624,
                648,
                672,
                696,
                720,
              ]
            }
            disabled={false}
            formatLabel={[Function]}
            help="Automatically resolve an issue if it hasn't been seen for this amount of time."
            hideErrorMessage={false}
            label="Auto resolve"
            max={720}
            min={0}
            name="resolveAge"
            required={false}
            snap={true}
            step={1}
          />
          <p>
            <small>
              <strong>
                Note: Enabling auto resolve will immediately resolve anything that has not been seen within this period of time. There is no undo!
              </strong>
            </small>
          </p>
        </div>
      </div>
      <div
        className="box"
      >
        <div
          className="box-header"
        >
          <h3>
            Data Privacy
          </h3>
        </div>
        <div
          className="box-content with-padding"
        >
          <BooleanField
            disabled={false}
            disabledReason="This option is enforced by your organization's settings and cannot be customized per-project."
            help="Enable server-side data scrubbing."
            hideErrorMessage={false}
            label="Data scrubber"
            name="dataScrubber"
            required={false}
            value={null}
          />
          <BooleanField
            disabled={false}
            disabledReason="This option is enforced by your organization's settings and cannot be customized per-project."
            help="Apply default scrubbers to prevent things like passwords and credit cards from being stored."
            hideErrorMessage={false}
            label="Use default scrubbers"
            name="dataScrubberDefaults"
            required={false}
            value={null}
          />
          <ListAsTextareaField
            disabled={false}
            help="Additional field names to match against when scrubbing data. Separate multiple entries with a newline."
            hideErrorMessage={false}
            label="Additional sensitive fields"
            name="sensitiveFields"
            placeholder="e.g. email"
            required={false}
          />
          <ListAsTextareaField
            disabled={false}
            help="Field names which data scrubbers should ignore. Separate multiple entries with a newline."
            hideErrorMessage={false}
            label="Safe fields"
            name="safeFields"
            placeholder="e.g. email"
            required={false}
          />
          <BooleanField
            disabled={false}
            disabledReason="This option is enforced by your organization's settings and cannot be customized per-project."
            help="Prevent IP addresses from being stored for new events."
            hideErrorMessage={false}
            label="Don't store IP Addresses"
            name="scrubIPAddresses"
            required={false}
            value={null}
          />
        </div>
      </div>
      <div
        className="box"
      >
        <div
          className="box-header"
        >
          <h3>
            Client Security
          </h3>
        </div>
        <div
          className="box-content with-padding"
        >
          <p>
            <span
              key="4"
            >
              <span
                key="0"
              >
                Configure origin URLs which Sentry should accept events from. This is used for communication with clients like 
              </span>
              <a
                href="https://github.com/getsentry/raven-js"
                key="1"
              >
                raven-js
              </a>
              <span
                key="2"
              >
                .
              </span>
            </span>
             
            <span
              key="6"
            >
              <span
                key="0"
              >
                This will restrict requests based on the 
              </span>
              <code
                key="1"
              >
                Origin
              </code>
              <span
                key="2"
              >
                 and 
              </span>
              <code
                key="3"
              >
                Referer
              </code>
              <span
                key="4"
              >
                 headers.
              </span>
            </span>
          </p>
          <ListAsTextareaField
            disabled={false}
            help="Separate multiple entries with a newline. Cannot be empty."
            hideErrorMessage={false}
            label="Allowed domains"
            name="allowedDomains"
            placeholder="e.g. https://example.com or example.com"
            required={false}
          />
          <BooleanField
            disabled={false}
            help="Allow Sentry to scrape missing JavaScript source context when possible."
            hideErrorMessage={false}
            label="Enable JavaScript source fetching"
            name="scrapeJavaScript"
            required={false}
          />
          <TextField
            disabled={false}
            help="Outbound requests matching Allowed Domains will have the header \\"{token_header}: {token}\\" appended."
            hideErrorMessage={false}
            label="Security token"
            name="securityToken"
            required={false}
          />
          <TextField
            disabled={false}
            help="Outbound requests matching Allowed Domains will have the header \\"{token_header}: {token}\\" appended."
            hideErrorMessage={false}
            label="Security token header"
            name="securityTokenHeader"
            placeholder="e.g. X-Sentry-Token"
            required={false}
          />
          <BooleanField
            disabled={false}
            help="Outbound requests will verify TLS (sometimes known as SSL) connections."
            hideErrorMessage={false}
            label="Verify TLS/SSL"
            name="verifySSL"
            required={false}
          />
        </div>
      </div>
      <div
        className="box"
      >
        <div
          className="box-header"
        >
          <h3>
            Remove Project
          </h3>
        </div>
        <div
          className="box-content with-padding"
        >
          <p>
            <a
              className="btn btn-danger pull-right"
              href="/org-slug/project-slug/settings/remove/"
            >
              Remove Project
            </a>
            Remove the 
            <strong>
              project-slug
            </strong>
             project and all related data.
            <br />
            Careful, this action cannot be undone.
          </p>
        </div>
      </div>
      <div
        className="box"
      >
        <div
          className="box-header"
        >
          <h3>
            Transfer Project
          </h3>
        </div>
        <div
          className="box-content with-padding"
        >
          <p>
            <a
              className="btn btn-danger pull-right"
              href="/org-slug/project-slug/settings/transfer/"
            >
              Transfer Project
            </a>
            Transfer the 
            <strong>
              project-slug
            </strong>
             project and all related data.
            <br />
            Careful, this action cannot be undone.
          </p>
        </div>
      </div>
    </ApiForm>
  </div>
</SideEffect(DocumentTitle)>
`;
